fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/typing/TypeRepository.java

```diff
@@ -471,20 +488,50 @@ public Builder setPackage(@Nonnull final String name) {
 
         @Nonnull
         public Builder addTypeIfNeeded(@Nonnull final Type type) {
-            if (!typeToNameMap.containsKey(type)) {
+            // Type.Null is special - it can only be nullable and has no non-nullable variant
+            if (type instanceof Type.Null) {
+                if (!nullableTypeToNameMap.containsKey(type)) {
+                    type.defineProtoType(this);
+                }
+                return this;
+            }
```

Comment:
I believe `Relation`s and `Any` are also special in that they also have a fixed nullability. (`Relation`s are always non-null, and `Any` is always nullable.)

It's a bit weird, as none of those types actually override `defineProtoType`, so they won't end up adding anything to the type repository anyway. But I suppose it doesn't hurt to make sure we cover for them. Potentially a method like `canonicalizeNullability(Type)` would be useful that returned (1) a non-nullable `Relation` and `None`, and (2) a nullable value for everything else. Then that would be the type we probed for.